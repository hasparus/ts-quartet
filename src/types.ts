export interface IContext {
  explanations: any[];
  pure: boolean;
  [key: string]: any;
}
export type Prepare = (ctx: IContext) => void;
export type HandleError = (valueId: string, ctxId: string) => string;
export interface IFunctionSchemaResult {
  prepare?: Prepare;
  check: (valueId: string, ctxId: string) => string;
  handleError?: HandleError;
  not?: (valudId: string, ctxId: string) => string;
}

export type FunctionSchema = () => IFunctionSchemaResult;

export interface IObjectSchema {
  "__quartet/rest-omit__"?: string[];
  [key: string]: Schema;
}
export interface IVariantSchema extends Array<Schema> {}
export type ConstantSchema =
  | undefined
  | null
  | boolean
  | number
  | string
  | symbol;
export type Schema =
  | FunctionSchema
  | ConstantSchema
  | IObjectSchema
  | IVariantSchema;

export type HandleSchemaHandler<T extends Schema, R> = (schema: T) => R;
export interface IHandleSchemaHandlers<R> {
  constant: HandleSchemaHandler<ConstantSchema, R>;
  function: HandleSchemaHandler<FunctionSchema, R>;
  object: HandleSchemaHandler<IObjectSchema, R>;
  objectRest: HandleSchemaHandler<IObjectSchema, R>;
  variant: HandleSchemaHandler<IVariantSchema, R>;
}

export interface ITest {
  test: (value: any) => boolean;
}

export type CustomFunction = ((value: any) => boolean) & {
  explanations?: any[];
  pure?: boolean;
};

export type CompilationResult = ((value: any) => boolean) & IContext;
export type TypedCompilationResult<T> = ((value: any) => value is T) & IContext;

export interface IMethods {
  and: (this: QuartetInstance, ...schemas: Schema[]) => FunctionSchema;
  arrayOf: (this: QuartetInstance, schema: Schema) => FunctionSchema;
  boolean: FunctionSchema;
  compileAnd: (<T = any>(this: QuartetInstance, ...schemas: Schema[]) => TypedCompilationResult<T>) &
    ((this: QuartetInstance, ...schemas: Schema[]) => CompilationResult);
  compileArrayOf: (<T = any>(this: QuartetInstance, schema: Schema) => TypedCompilationResult<T[]>) &
    ((this: QuartetInstance, schema: Schema) => CompilationResult);
  custom: (this: QuartetInstance, check: CustomFunction, explanation?: any) => FunctionSchema;
  finite: FunctionSchema;
  function: FunctionSchema;
  max: (this: QuartetInstance, maxValue: number, exclusive?: boolean) => FunctionSchema;
  maxLength: (this: QuartetInstance, maxLength: number, exclusive?: boolean) => FunctionSchema;
  min: (this: QuartetInstance, minValue: number, exclusive?: boolean) => FunctionSchema;
  minLength: (this: QuartetInstance, minLength: number, exclusive?: boolean) => FunctionSchema;
  negative: FunctionSchema;
  not: (this: QuartetInstance, schema: Schema) => FunctionSchema;
  number: FunctionSchema;
  positive: FunctionSchema;
  rest: "__quartet/rest__";
  restOmit: "__quartet/rest-omit__";
  safeInteger: FunctionSchema;
  string: FunctionSchema;
  symbol: FunctionSchema;
  test: (this: QuartetInstance, test: ITest) => FunctionSchema;
}

export type PureCompile = (schema: Schema) => CompilationResult
export type ToContext = (prefix: string | number, value: any, isUniq?: boolean) => [string, Prepare]

interface InnerMethods {
  pureCompile: PureCompile
  toContext: ToContext
  clearContextCounters: () => void
}

export type QuartetInstance = IMethods & InnerMethods &
  (<T>(schema: Schema) => TypedCompilationResult<T>) &
  ((schema: Schema) => CompilationResult);
