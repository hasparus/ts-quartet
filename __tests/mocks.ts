import {
  FunctionSchema,
  IVariantSchema,
  Schema,
  IObjectSchema
} from "../types";
import { v } from "../index";

export const primitives: any[] = [
  undefined,
  null,
  true,
  false,
  0,
  1,
  NaN,
  Infinity,
  -Infinity,
  "",
  "Andrew",
  Symbol.for("test")
];

export const funcSchema: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`
});
export const funcSchemaWithNot: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 2 !== 0)`
});
export const funcSchemaWithHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});
export const funcSchemaWithNotHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 2 !== 0)`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});
export const funcSchemaWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithNotWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  not: (value, ctx) =>
    `(typeof ${value} !== 'number' || ${value} % ${ctx}.divide !== 0)`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithHandleErrorWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithNotHandleErrorWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`,
  not: (value, ctx) =>
    `(typeof ${value} !== 'number' || ${value} % ${ctx}.divider !== 0)`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});

export const funcSchemas: FunctionSchema[] = [
  funcSchema,
  funcSchemaWithNot,
  funcSchemaWithHandleError,
  funcSchemaWithNotHandleError,
  funcSchemaWithPrepare,
  funcSchemaWithNotWithPrepare,
  funcSchemaWithHandleErrorWithPrepare,
  funcSchemaWithNotHandleErrorWithPrepare
];

const has = (obj: any, key: string | number) => {
  return Object.prototype.hasOwnProperty.call(obj, key);
};

export function* objects(
  level: number = 0,
  maxLevel: number = 2
): Iterable<object | any[]> {
  if (level > maxLevel) return;
  yield {};
  yield [];
  for (const obj of objects(level + 1, maxLevel)) {
    if (has(obj, "push")) {
      for (const primitive of primitives) {
        yield [...(obj as any), primitive];
      }
      for (const o of objects(level + 1, maxLevel)) {
        yield [...(obj as any), o];
      }
    } else {
      for (const primitive of primitives) {
        if (!has(obj, "even")) {
          yield { ...obj, even: primitive };
        }
        if (!has(obj, "another")) {
          yield { ...obj, another: primitive };
        }
      }
      for (const o of objects(level + 1, maxLevel)) {
        if (!has(obj, "even")) {
          yield { ...obj, even: o };
        }
        if (!has(obj, "another")) {
          yield { ...obj, another: o };
        }
        if (!has(obj, "nested")) {
          yield { ...obj, nested: o };
        }
      }
    }
  }
}

export function* variants(
  level: number = 0,
  maxLevel: number = 2
): Iterable<IVariantSchema> {
  if (level > maxLevel) {
    return;
  }
  yield [];
  for (const variant of variants(level + 1)) {
    for (const primitive of primitives) {
      yield [...variant, primitive];
    }
    for (const funcSchema of funcSchemas) {
      yield [...variant, funcSchema];
    }
    for (const obj of objectSchemas(level + 1, maxLevel)) {
      yield [...variant, obj];
    }
    for (const variant2 of variants(level + 1, maxLevel)) {
      yield [...variant, variant2];
    }
  }
}

export function* objectSchemas(
  level: number = 0,
  maxLevel: number = 2
): Iterable<IObjectSchema> {
  if (level > maxLevel) return;
  yield {};
  for (const obj of objectSchemas(level + 1, maxLevel)) {
    if (!has(obj, "even")) {
      for (const primitive of primitives) {
        yield { ...obj, even: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, even: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, even: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), even: obj };
      }
    }
    if (!has(obj, "nested")) {
      for (const primitive of primitives) {
        yield { ...obj, nested: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, nested: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, nested: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), nested: obj };
      }
    }
    if (!has(obj, v.rest)) {
      for (const primitive of primitives) {
        yield { ...obj, [v.rest]: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, [v.rest]: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, [v.rest]: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), [v.rest]: obj };
      }
    }
  }
}

export function* schemas(maxLevel: number): Iterable<Schema> {
  yield* primitives[Symbol.iterator]();
  yield* funcSchemas[Symbol.iterator]();
  yield* objectSchemas(0, maxLevel);
  yield* variants(0, maxLevel);
}

export function* values(maxLevel: number) {
  yield* primitives[Symbol.iterator]();
  yield* objects(0, maxLevel);
}
