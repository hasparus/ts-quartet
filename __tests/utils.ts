import { getDescription } from "./getDescription";

export function snapshot(validator: any) {
  const description = getDescription(validator);

  expect(description).toMatchSnapshot();
}
export function getExplanation(
  validator: { (arg0: any): void; explanations: any },
  value: any
) {
  validator(value);
  return validator.explanations;
}
export function isValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (!valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(true);
}
export function isNotValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(false);
}

export function tables(validator: any, valids: any[], invalids: any[]) {
  for (const valid of valids) {
    isValid(validator, valid);
  }
  for (const invalid of invalids) {
    isNotValid(validator, invalid);
  }
}
