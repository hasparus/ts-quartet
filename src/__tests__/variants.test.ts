import { v } from "../index";
import { snapshot, tables, puretables } from "./utils";
import { funcSchemaWithNot, primitives } from "./mocks";
import { FunctionSchema } from "../types";

const funcSchemaWithNotHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 3 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 3 !== 0)`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});

describe("variants", () => {
  test("01. v([])", () => {
    const validator = v([]);
    expect(validator.pure).toBe(true);
    snapshot(validator);
    puretables(validator, [], [...primitives, {}, []]);
  });
  test("02. v([pureFunc])", () => {
    const validator = v([funcSchemaWithNot]);
    expect(validator.pure).toBe(true);
    snapshot(validator);
    puretables(
      validator,
      [2, 4, 6, 8],
      [
        ...primitives.filter(e => typeof e !== "number" || e % 2 !== 0),
        1,
        3,
        5,
        7,
        {},
        []
      ]
    );
  });
  test("03. v([pureFunc, is3k])", () => {
    const validator = v([funcSchemaWithNot, funcSchemaWithNotHandleError]);
    expect(validator.pure).toBe(false);
    snapshot(validator);
    tables(validator, [2, 4, 3, 6, 8, 9, 0], [[1, [1]]]);
  });
  test("04. v([pureFunc, pureFunc])", () => {
    const validator = v([funcSchemaWithNot, funcSchemaWithNot]);
    expect(validator.pure).toBe(true);
    snapshot(validator);
    puretables(
      validator,
      [2, 4, 6, 8],
      [
        ...primitives.filter(e => typeof e !== "number" || e % 2 !== 0),
        1,
        3,
        5,
        7,
        {},
        []
      ]
    );
  });
  test("05. v([1,2,3,4,5])", () => {
    const validator = v([1, 2, 3, 4, 5]);
    expect(validator.pure).toBe(true);
    snapshot(validator);
    puretables(
      validator,
      [1, 2, 3, 4, 5],
      [
        "1",
        "2",
        "3",
        "4",
        "5",
        ...primitives.filter(e => typeof e !== "number" || e < 1 || e > 5)
      ]
    );
  });
  test('06. v(["A", "B", Symbol.for("test")])', () => {
    const validator = v(["A", "B", Symbol.for("variant")]);
    expect(validator.pure).toBe(true);
    snapshot(validator);
    puretables(
      validator,
      ["A", "B", Symbol.for("variant")],
      [1, 2, 3, 4, 5, ...primitives]
    );
  });
  test('07. v(["A", "B", 1,2,3, isEven])', () => {
    const validator = v(["A", "B", 1, 2, 3, funcSchemaWithNot]);
    expect(validator.pure).toBe(true);
    snapshot(validator);
    puretables(
      validator,
      ["A", "B", 1, 2, 3, 4, 6, 8, 10],
      [5, ...primitives.filter(e => typeof e !== "number")]
    );
  });
});
