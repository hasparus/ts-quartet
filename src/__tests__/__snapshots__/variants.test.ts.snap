// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variants 01. v([]) 1`] = `
Object {
  "_": "function () { return false; }",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`variants 02. v([pureFunc]) 1`] = `
Object {
  "_": "function validator(value) {
  return typeof value === 'number' && value % 2 === 0
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`variants 03. v([pureFunc, is3k]) 1`] = `
Object {
  "_": "function validator(value) {
  validator.explanations = []
  if (typeof value === 'number' && value % 2 === 0) return true;
  if (typeof value === 'number' && value % 3 === 0) return true;
  validator.explanations.push(value)
  return false
}",
  "explanations": Array [],
  "pure": false,
}
`;

exports[`variants 04. v([pureFunc, pureFunc]) 1`] = `
Object {
  "_": "function validator(value) {
  if (typeof value === 'number' && value % 2 === 0) return true;
  if (typeof value === 'number' && value % 2 === 0) return true;
  return false
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`variants 05. v([1,2,3,4,5]) 1`] = `
Object {
  "_": "function validator(value) {
  if (value === 1) return true;
  if (value === 2) return true;
  if (value === 3) return true;
  if (value === 4) return true;
  if (value === 5) return true;
  return false
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`variants 06. v(["A", "B", Symbol.for("test")]) 1`] = `
Object {
  "_": "function validator(value) {
  if (validator.__validValuesDict[value] === true) return true
  return false
}",
  "__validValuesDict": Object {
    "A": true,
    "B": true,
    Symbol(variant): true,
  },
  "explanations": Array [],
  "pure": true,
}
`;

exports[`variants 07. v(["A", "B", 1,2,3, isEven]) 1`] = `
Object {
  "_": "function validator(value) {
  if (validator.__validValuesDict[value] === true) return true
  if (value === 1) return true;
  if (value === 2) return true;
  if (value === 3) return true;
  if (typeof value === 'number' && value % 2 === 0) return true;
  return false
}",
  "__validValuesDict": Object {
    "A": true,
    "B": true,
  },
  "explanations": Array [],
  "pure": true,
}
`;
